torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -57,7 +54,42 @@ def _extract(file: pathlib.Path) -> None:
     def _decompress(file: pathlib.Path) -> None:
         _decompress(str(file), remove_finished=True)
 
-    def _loader(self, path: pathlib.Path) -> IterDataPipe[Tuple[str, IO]]:
+    @classmethod
+    def from_http(cls, url: str, *, file_name: Optional[str] = None, **kwargs: Any) -> OnlineResource:
+        return cls(url, file_name=file_name or pathlib.Path(urlparse(url).path).name, **kwargs)
+
+    @classmethod
+    def from_gdrive(cls, id: str, **kwargs: Any) -> OnlineResource:
+        return cls(f"https://drive.google.com/uc?export=download&id={id}", **kwargs)
+
+    def download(self, root: Union[str, pathlib.Path], *, skip_integrity_check: bool = False) -> pathlib.Path:
+        root = pathlib.Path(root).expanduser().resolve()
+        root.mkdir(parents=True, exist_ok=True)
+        file = root / self.file_name
+
+        if not file.exists():
+            dp = IterableWrapper([self.url])
+            dp = OnlineReader(dp)
+            stream = list(dp)[0][1]
+
+            with open(file, "wb") as fh, tqdm() as progress_bar:
+                for chunk in iter(lambda: stream.read(1024 * 1024), b""):  # type: ignore[no-any-return]
+                    # filter out keep-alive new chunks
+                    if not chunk:
+                        continue
+
+                    fh.write(chunk)
+                    progress_bar.update(len(chunk))
```

Comment:
There a two reasons I didn't do this with datapipes:

1. We should probably have a progress bar for downloads, since they usually are quite large and a long running process without output is always suspicious. By importing `tqdm` from `torch.hub` we get vanilla `tqdm` if it is installed and a minimal port if it is not.
2. AFAIK, `torchdata` has no builtin option to read chunks from a stream. We need that, because we otherwise load the whole file into memory before saving it.

cc @ejguan @NivekT

Reply:
1. I agree having a progress bar for download is a good idea. Maybe `torchdata` can have a dependency for `tqdm`.
2. I suppose adding an option to return `r.iter_content()` instead of `r.raw` would work?

Reply:
1. Apart from [`tqdm`](https://github.com/tqdm/tqdm), [`rich`](https://github.com/Textualize/rich) also added support for [progress bars](https://rich.readthedocs.io/en/stable/progress.html) recently. That is what newer versions of `pip` are using. Just leaving this here to asses the options before deciding to depend on `tqdm`.
2. I would avoid having a flag the changes the output types. Having something like a `ChunkReader` datapipe can be useful in general not just for HTTP requests, right?

    ```py
    from torchdata.datapipes.iter import IterDataPipe
    
    
    class ChunkReader(IterDataPipe):
        def __init__(self, datapipe, *, chunk_size=32 * 1024 * 1024):
            self.datapipe = datapipe
            self.chunk_size = chunk_size
    
        def __iter__(self):
            for path, stream in self.datapipe:
                for chunk in iter(lambda: stream.read(self.chunk_size), b""):
                    # filter out keep-alive new chunks from HTTP streams
                    if not chunk:
                        continue
    
                    yield path, chunk
    ```

    A utility like that is needed anyway for `Saver` anyway. AFAIK, there is currently no builtin functionality to read data from a stream, correct? I'm guessing that is why the [`Saver` example](https://github.com/pytorch/data/blob/12cfaf8899b1337981cd4edf9deef127f925f1bd/torchdata/datapipes/iter/util/saver.py#L34-L36) uses `bytes` as input.

If there is more to discuss, let's migrate this to a proper issue on the `torchdata` repository.

Reply:
I will open an issue in `torchdata` to get more thoughts from the community and see if there is anything that we may be missing.

IIUC, since `_get_response_from_http` returns `r.raw`, rather than say `r.iter_content()`, it loads the whole response into memory at once. Is that correct?

https://github.com/pytorch/data/blob/12cfaf8899b1337981cd4edf9deef127f925f1bd/torchdata/datapipes/iter/load/online.py#L46

Reply:
> I will open an issue in `torchdata` to get more thoughts from the community and see if there is anything that we may be missing.

See pytorch/data#439. It seems we missed `StreamReader`, which does exactly that. Still, could you open an issue regarding the progress bar?

> IIUC, since `_get_response_from_http` returns `r.raw`, rather than say `r.iter_content()`, it loads the whole response into memory at once. Is that correct?

`r.raw` just gives you the stream and is somewhat like `BinaryIO`, whereas `r.iter_content()` reads chunks from the stream and thus is a `Iterator[bytes]`. IMO returning `r.raw` is the right choice, since attaching a `StreamReader` to a `HTTPReader` will give us the `iter_content()` behavior.